pkg/segment/reader/record/recordreader.go

```diff
@@ -36,6 +36,171 @@ import (
+	case "_index":
+		return readIndexForRRCs(rrcs)
+	default:
+		return readUserDefinedColForRRCs(segKey, rrcs, cname, qid)
+	}
+}
+
+func readTimestampForRRCs(rrcs []*utils.RecordResultContainer) ([]utils.CValueEnclosure, error) {
+	result := make([]utils.CValueEnclosure, len(rrcs))
+	for i, rrc := range rrcs {
+		result[i] = utils.CValueEnclosure{
+			Dtype: utils.SS_DT_UNSIGNED_NUM,
+			CVal:  rrc.TimeStamp,
+		}
+	}
+
+	return result, nil
+}
+
+func readIndexForRRCs(rrcs []*utils.RecordResultContainer) ([]utils.CValueEnclosure, error) {
+	result := make([]utils.CValueEnclosure, len(rrcs))
+	for i, rrc := range rrcs {
+		result[i] = utils.CValueEnclosure{
+			Dtype: utils.SS_DT_STRING,
+			CVal:  rrc.VirtualTableName,
+		}
+	}
+
+	return result, nil
+}
+
+// All the RRCs must belong to the same segment.
+func readUserDefinedColForRRCs(segKey string, rrcs []*utils.RecordResultContainer,
+	cname string, qid uint64) ([]utils.CValueEnclosure, error) {
+
+	if len(rrcs) == 0 {
+		return nil, nil
+	}
+
+	err := fileutils.GLOBAL_FD_LIMITER.TryAcquireWithBackoff(1, 10, "readUserDefinedColForRRCs")
+	if err != nil {
+		log.Errorf("qid=%v, readUserDefinedColForRRCs failed to get lock for opening 1 file; err=%v", qid, err)
+		return nil, err
+	}
+	defer fileutils.GLOBAL_FD_LIMITER.Release(1)
+
+	blockMetadata, err := writer.GetBlockSearchInfoForKey(segKey)
+	if err != nil {
+		log.Errorf("qid=%v, readUserDefinedColForRRCs: failed to get block metadata for segKey %s; err=%v", qid, segKey, err)
+		return nil, err
+	}
+
+	blockSummary, err := writer.GetBlockSummaryForKey(segKey)
+	if err != nil {
+		log.Errorf("qid=%v, readUserDefinedColForRRCs: failed to get block summary for segKey %s; err=%v", qid, segKey, err)
+		return nil, err
+	}
+
+	consistentCValLen := map[string]uint32{cname: utils.INCONSISTENT_CVAL_SIZE} // TODO: use correct value
+	sharedReader, err := segread.InitSharedMultiColumnReaders(segKey, map[string]bool{cname: true},
+		blockMetadata, blockSummary, 1, consistentCValLen, qid)
+	if err != nil {
+		log.Errorf("qid=%v, readUserDefinedColForRRCs: failed to initialize shared readers for segkey %v; err=%v", qid, segKey, err)
+		return nil, err
+	}
+	defer sharedReader.Close()
+	multiReader := sharedReader.MultiColReaders[0]
+
+	batchingFunc := func(rrc *utils.RecordResultContainer) uint16 {
+		return rrc.BlockNum
+	}
+	batchKeyLess := toputils.NewOptionWithValue(func(blockNum1, blockNum2 uint16) bool {
+		// We want to read the file in order, so read the blocks in order.
+		return blockNum1 < blockNum2
+	})
+	operation := func(rrcsInBatch []*utils.RecordResultContainer) []utils.CValueEnclosure {
+		if len(rrcsInBatch) == 0 {
+			return nil
+		}
+
+		return handleBlock(multiReader, rrcsInBatch[0].BlockNum, rrcsInBatch, qid)
+	}
+
+	enclosures := toputils.BatchProcess(rrcs, batchingFunc, batchKeyLess, operation)
+	return enclosures, nil
+}
+
+func handleBlock(multiReader *segread.MultiColSegmentReader, blockIdx uint16,
```

Comment:
nit: should we call this `blockNum` as we have been referring to it that way in most of the code base ? 
but also ok with keeping it `blockIdx`

Reply:
I'll change that in a different PR. It was called blockIdx other places in this file, but I agree blockNum is better